src/lem/store_core.rs

```diff
@@ -0,0 +1,398 @@
+                    let [a, b] = self.expect_tuple2(idx);
+                    let a_digest = self.hash_ptr_val_unsafe(a.val());
+                    let b_digest = self.hash_ptr_val_unsafe(b.val());
+                    let a = GPtr::new(*a.tag(), a_digest);
+                    let b = GPtr::new(*b.tag(), b_digest);
+                    self.hasher.hash_ptrs(vec![a, b])
+                }
+                IVal::Tuple3(idx) => {
+                    let [a, b, c] = self.expect_tuple3(idx);
+                    let a_digest = self.hash_ptr_val_unsafe(a.val());
+                    let b_digest = self.hash_ptr_val_unsafe(b.val());
+                    let c_digest = self.hash_ptr_val_unsafe(c.val());
+                    let a = GPtr::new(*a.tag(), a_digest);
+                    let b = GPtr::new(*b.tag(), b_digest);
+                    let c = GPtr::new(*c.tag(), c_digest);
+                    self.hasher.hash_ptrs(vec![a, b, c])
+                }
+                IVal::Tuple4(idx) => {
+                    let [a, b, c, d] = self.expect_tuple4(idx);
+                    let a_digest = self.hash_ptr_val_unsafe(a.val());
+                    let b_digest = self.hash_ptr_val_unsafe(b.val());
+                    let c_digest = self.hash_ptr_val_unsafe(c.val());
+                    let d_digest = self.hash_ptr_val_unsafe(d.val());
+                    let a = GPtr::new(*a.tag(), a_digest);
+                    let b = GPtr::new(*b.tag(), b_digest);
+                    let c = GPtr::new(*c.tag(), c_digest);
+                    let d = GPtr::new(*d.tag(), d_digest);
+                    self.hasher.hash_ptrs(vec![a, b, c, d])
+                }
+                IVal::Compact(idx) => {
+                    let [a, b, c] = self.expect_compact(idx);
+                    let a_digest = self.hash_ptr_val_unsafe(a.val());
+                    let b_digest = self.hash_ptr_val_unsafe(b.val());
+                    let c_digest = self.hash_ptr_val_unsafe(c.val());
+                    self.hasher
+                        .hash_compact(a_digest, *b.tag(), b_digest, c_digest)
+                }
+            };
+            self.z_cache.insert(*ptr_val, Box::new(digest));
+            self.inverse_z_cache.insert(digest, Box::new(*ptr_val));
+            digest
+        }
+    }
+
+    /// Hashes `IVal`s in parallel, consuming chunks of length 256, which is a
+    /// reasonably safe limit. The danger of longer chunks is that the rightmost
+    /// pointers are the ones which are more likely to reach the recursion depth
+    /// limit in `hash_ptr_val_unsafe`. So we move in smaller chunks from left to
+    /// right, populating the `z_cache`, which can rescue `hash_ptr_val_unsafe`
+    /// from deep recursions
+    fn hydrate_z_cache_with_ptr_vals(&self, ptrs: &[&IVal]) {
+        ptrs.chunks(256).for_each(|chunk| {
+            chunk.par_iter().for_each(|ptr| {
+                self.hash_ptr_val_unsafe(ptr);
+            });
+        });
+    }
```

Comment:
We also need to think about this, though: we want to avoid unnecessary hashing, and when we are no longer using all hashes for internal structure, that means there may well be pointers in the store that never need to be hashed. *One* strategy for proofs may involve ensuring that the store contains no superfluous hashes — so all those it does contain can be blindly transcribed into the trace.

I haven't thought this through completely, other than to note that the complete hydration method may somewhat contradict this approach.

Reply:
> I think it would be better to have some explicit topological sorting mechanism

That's pretty much what `hash_ptr_val` does before calling `hash_ptr_val_unsafe`. So we can just call `hash_ptr` on specific pointers if we don't want to call `hydrate_z_cache`